replace get<T>.tie(v,e) with get(v)

jkeiser opened the pull request:
This adds a get(v) method that acts as ergonomic shorthand for get<type>().tie(v, error). It returns the error code so you can use it in contorl flow. I like it even better than `auto [v,error]` as well, in most cases. It simplifies things considerably.

### Comparison

* New version:
  ```c++
  int64_t val;
  if ((error = obj["val"].get(val))) { cerr << error << endl; return; }
  cout << val << endl;
  ```
* Old version with auto (C++14): *this feels like it buries the lead (the fact that we're getting an int64_t out)*
  ```c++
  auto [val, error] = obj["val"].get<int64_t>();
  if (error) { cerr << error << endl; return; }
  cout << val << endl;
  ```
* Old version with tie (C++11): *this is verbose and not control flow friendly*
  ```c++
  int64_t val;
  obj["val"].get<int64_t>().tie(val, error);
  if (error) { cerr << error << endl; return; }
  cout << val << endl;
  ```
* New version without if: *For comparison, if you don't like stuff in your if statements :)*
  ```c++
  int64_t val;
  error = obj["val"].get(val);
  if (error) { cerr << error << endl; return; }
  cout << val << endl;
  ```

### Changes

* You don't have to write the type name twice anymore.
* You can include the call in control flow (as an if or while condition) if you so desire.
* It's even better than auto in many cases because it's easier see the type of your variable when you declare it (since it's prominent on the left where you are usually looking).
* I added it to simdjson_result and it feels consistent throughout the library.
* I had it stop assigning to value if there is an error, and uninitialized warnings revealed a few places where we were using value even when there was an error! I would like to leave this for developer safety--make it impossible to do this without a warning.
  - To do this, I added SIMDJSON_UNREACHABLE and SIMDJSON_ASSUME macros to use in perf tests.

## Example

I've converted the README, tools, benchmarks, and a lot of the examples to it.

jkeiser commented:
This fixes #606.

jkeiser commented:
This also significantly mitigates #949: in most cases, if you're using error codes, you don't need to use templates. There are still cases, though.

jkeiser commented:
One other form I kind of wanted to do was a single-value form:

```c++
template<T>
WARN_UNUSED error_code get(T &value);
```

But I didn't, because the other get() returns true on error. That would make it so when you are using the two-argument form, `if (get(v,err))` is how you check for success, but when you use the one-argument form, `if (!get(v))` is how you check for success. Having two functions with the same name but opposite result values seemed like a bridge too far.

lemire commented:
Let me start with a stupid question... why not do...

```C++
if(error = get(val)) {
  // I have the error "error"
}
```

(To avoid warnings, you may need to do...
```C++
if((error = get(val))) {
  // I have the error "error"
}
```
)



In C++17, you can even do...
```C++
if(error_code error;  
     error = get(val)) {
  // I have the error "error"
}
```

so you don't even need to dirty you code with the error variable.

lemire commented:
One use might be this...

```C++

struct mystruct {
  double x;
  int64_t y;
  std:;string_view z;
};
mystruct s;
if(doc.get(s.x,error) && doc.get(s.y,error) && doc.get(s.z,error)) {
 // got struct!
}
```

This could still work as...

```C++

struct mystruct {
  double x;
  int64_t y;
  std:;string_view z;
};
mystruct s;
if(error = doc.get(s.x) && doc.get(s.y) && doc.get(s.z)) {
 // got struct!
}
```

but extra code might be needed.

lemire commented:
I see that a follow-up comment answers my first question.

lemire commented on benchmark/bench_dom_api.cpp:
```diff
@@ -18,17 +18,15 @@ static void numbers_scan(State& state) {
-  parser.load(NUMBERS_JSON).get<dom::array>().tie(arr, error);
-  if(error) {
-    cerr << "could not read " << NUMBERS_JSON << " as an array" << endl;
+  if (!parser.load(NUMBERS_JSON).get(arr, error)) {
```
If we went the other way (return error), we would have...

```C++
if (error  = parser.load(NUMBERS_JSON).get(arr)) {
}
```

instead of 


```C++
if (!parser.load(NUMBERS_JSON).get(arr, error)) {
}
```

jkeiser replied:
I liked this a lot, so I changed it to this :)

You do need double parentheses to deal with a warning about = in conditions:

```c++
if ((error = parser.load(NUMBERS_JSON).get(arr))) {
}
```

I find it just as readable. More so in some ways: the double paren can serve as a kind of marker that you're seeing one of these error = function() constructions.

And I really like the technique of assigning the error. Otherwise you end up with !function() being success and function() meaning an error, which is just not intuitive.

lemire commented on benchmark/bench_dom_api.cpp:
```diff
@@ -365,11 +356,12 @@ BENCHMARK(twitter_image_sizes);
-    if (error) { return; }
+    uint64_t value;
+    if (!doc["search_metadata"]["count"].get(value, error)) { return; }
     if (value != 100) { return; }
```
Let us compare here... We had....

```C++
auto [value, error] = doc["search_metadata"]["count"].get<uint64_t>();
if (error) { return; }
```

and we are moving to...

```C++
uint64_t value;
if (!doc["search_metadata"]["count"].get(value, error)) { return; }
```

jkeiser replied:
Re-comparing, we had:

```c++
auto [value, error] = doc["search_metadata"]["count"].get<uint64_t>();
if (error) { return; }
```

And are now moving to:

```c++
uint64_t value;
if ((error = doc["search_metadata"]["count"].get(value)) { return; }
```

Or, if you don't like stuff in your if statement:

```c++
uint64_t value;
auto error = doc["search_metadata"]["count"].get(value);
if (error) { return; }
```

It's kind of nice to be able to read the types of important variables like uint64_t on the left hand side when you are scanning. I am finding myself turning against the auto [value, error] syntax in favor of this :) What are your thoughts?

lemire commented on include/simdjson/dom/element.h:
```diff
@@ -206,6 +206,63 @@ class element {
+   * @returns true if the value was set, or false if there wsa an error.
+   */
+  template<typename T>
+  inline bool get(T &value, error_code &error) const noexcept;
```
The returned value is not WARN_UNUSED which I think is good.

jkeiser replied:
On purpose :) Since we're not doing the redundant bool thing anymore, the new get() is `WARN_UNUSED error_code get(T&)`.

jkeiser replied:
i.e. I changed it after your suggestions :)

lemire commented:
There is a lot of `if(!...)` which I don't like. it is confusing.

I find this construction easier to reason about...

```C++
int64_t val;
obj["val"].get(val, error);
if (error) { cerr << error << endl; }
```


It is close to what we had.... but better...


```C++
int64_t val;
obj["val"].tie<int64_t>().get(val, error);
if (error) { cerr << error << endl; }
```

lemire commented on tests/readme_examples_noexceptions.cpp:
```diff
@@ -35,87 +35,72 @@ void basics_error_3() {
+
+      // Casting a JSON element to an integer
+      uint64_t year;
+      if (!car["year"].get(year, error)) { cerr << error << endl; exit(1); }
```
My brain hurts when parsing...

```C++
if (!car["year"].get(year, error)) { cerr << error << endl; exit(1); }
```

I understand it at some level....

(By the way, in C++, you can type `not` instead of `!`.)

```C++
if (not car["year"].get(year, error)) { cerr << error << endl; exit(1); }
```

Already, to me, it looks better...

Anything to get close to...



```
if not get year from car[year]
```

but compared to English, it is not in order....

jkeiser commented:
@lemire let us see what happens if we *remove* the construction I just added, and replace it with only `WARN_UNUSED error_code get<T>(T&)`. That would remove the confusion of negation, and I hate out parameters anyway ...

(WARN_UNUSED because now the result is the only way to find out there's an error.)

jkeiser commented:
@lemire I removed get(v,e) and replaced it with e = get(v). I also added it to simdjson_result. I really like it a lot, to the point where I think we can stop recommending auto [v, error] = as a standard practice. Still need it in for loops for key/value pairs, but I think that's it!

lemire commented on include/simdjson/dom/element.h:
```diff
@@ -473,51 +472,51 @@ struct simdjson_result<dom::element> : public internal::simdjson_result_base<dom
   really_inline simdjson_result(error_code error) noexcept; ///< @private
 
-  inline simdjson_result<dom::element_type> type() const noexcept;
+  really_inline simdjson_result<dom::element_type> type() const noexcept;
```
Interesting that we did not have really_inline until now. I wonder whether this had performance impacts.

jkeiser replied:
I did this more as a consistency thing in a lot of cases: we are still using just inline in a lot of places we don't need to.

lemire commented:
I think you are getting warnings because...

```C++
if (error = size.value["h"].get(height)) { return; }
```

looks like an error so one has to do...
```C++
if ((error = size.value["h"].get(height))) { return; }
```

lemire commented:
> I really like it a lot

It also looks good to me and if you did not just do that to humour me, if you really like it, then it is a good sign.

lemire commented:
I stated it earlier, but with C++17, you can do...

```C++
if(error_code e; e = doc.get(val)) {
}
```

It has the benefit of isolating the declaration.

jkeiser commented:
> > I really like it a lot
> 
> It also looks good to me and if you did not just do that to humour me, if you really like it, then it is a good sign.

Not at all! I actually kind of want to deprecate tie() now. The only place you might need to use auto [a,b] is in a foreach loop over objects or document_stream, and tie won't work there anyway.

I'll suggest that and update documentation to stop suggesting auto [x,y] (except fields) in another PR.

jkeiser commented:
And I LOVE the idea that the best syntax for using simdjson is C++11 compliant, so you don't have to confuse people with alternative syntax :)

jkeiser commented:
Oh, the other thing I like about this over auto: it completely eliminates the aliasing warnings auto generates while still being petty readable.

lemire commented:
You have some kind of funny conflict with our main branch which causes failures?

We could deprecate tie.
